modules/crm_tools/src/Plugin/Field/FieldType/StatusLog.php

```diff
@@ -0,0 +1,168 @@
+      ->setLabel(new TranslatableMarkup('Previous Status'));
+
+    $properties['timestamp'] = DataDefinition::create('timestamp')
+      ->setLabel(new TranslatableMarkup('Changed timestamp'))
+      ->setRequired(TRUE);
+
+    $properties['uid'] = DataDefinition::create('integer')
+      ->setLabel(new TranslatableMarkup('User responsible for change'));
+
+    return $properties;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function schema(FieldStorageDefinitionInterface $field_definition) {
+    return [
+      'columns' => [
+        'value' => [
+          'type' => 'varchar',
+          'length' => (int) $field_definition->getSetting('max_length'),
+        ],
+        'previous' => [
+          'type' => 'varchar',
+          'length' => (int) $field_definition->getSetting('max_length'),
+        ],
+        'timestamp' => [
+          'type' => 'int',
+        ],
+        'uid' => [
+          'type' => 'int',
+        ],
+      ],
+      'indexes' => [
+        'value' => ['value'],
+        'previous' => ['previous'],
+      ],
+    ];
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getConstraints() {
+    $constraints = parent::getConstraints();
+
+    if ($max_length = $this->getSetting('max_length')) {
+      $constraint_manager = \Drupal::typedDataManager()->getValidationConstraintManager();
+      $constraints[] = $constraint_manager->create('ComplexData', [
+        'value' => [
+          'Length' => [
+            'max' => $max_length,
+            'maxMessage' => t('%name: may not be longer than @max characters.', [
+              '%name' => $this->getFieldDefinition()->getLabel(),
+              '@max' => $max_length
+            ]),
+          ],
+        ],
+      ]);
+    }
+
+    return $constraints;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function applyDefaultValue($notify = TRUE) {
+    parent::applyDefaultValue($notify);
+    return $this;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function generateSampleValue(FieldDefinitionInterface $field_definition) {
+    $random = new Random();
+
+    $values['value'] = $random->word(mt_rand(1, $field_definition->getSetting('max_length')));
+    $values['previous'] = $random->word(mt_rand(1, $field_definition->getSetting('max_length')));
+    $values['uid'] = mt_rand(0, 1);
+    $values['timestamp'] = mt_rand(1262055681, 1262055681);
+
+    return $values;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function storageSettingsForm(array &$form, FormStateInterface $form_state, $has_data) {
+    $elements = [];
+
+    $elements['max_length'] = [
+      '#type' => 'number',
+      '#title' => t('Maximum length'),
+      '#default_value' => $this->getSetting('max_length'),
+      '#required' => TRUE,
+      '#description' => t('The maximum length of the field in characters.'),
+      '#min' => 1,
+      '#disabled' => $has_data,
+    ];
+
+    // @todo Make this field a select field of existing fields on entity.
```

Comment:
When we do this, we need to filter the fields by appropriate ones. My suggestion would be ones where there is a main value and the main value is a string... We may want to put some similar validation somewhere else for programmatic fields, but not exactly sure where that should live...